File: functional/scala/nanocode.scala
//> using scala 3.3
//> using dep com.lihaoyi::requests:0.8.0
//> using dep com.lihaoyi::ujson:3.1.3
// nanocode - minimal claude code alternative (Scala 3)
import scala.io.StdIn, scala.util.*, java.io.File, scala.sys.process.*

val KEY = sys.env("ANTHROPIC_API_KEY")
val MODEL = sys.env.getOrElse("MODEL", "claude-sonnet-4-20250514")
val (R,B,D,C,G,BL) = ("\u001b[0m","\u001b[1m","\u001b[2m","\u001b[36m","\u001b[32m","\u001b[34m")

def runTool(name: String, input: ujson.Value): String = Try(name match
  case "read" => scala.io.Source.fromFile(input("path").str).getLines.zipWithIndex.map((l,i) => s"${i+1}| $l").mkString("\n")
  case "write" => java.nio.file.Files.writeString(java.nio.file.Path.of(input("path").str), input("content").str); "ok"
  case "edit" => val f = new File(input("path").str); val t = scala.io.Source.fromFile(f).mkString
    if !t.contains(input("old").str) then "error: not found" else { java.nio.file.Files.writeString(f.toPath, t.replaceFirst(java.util.regex.Pattern.quote(input("old").str), input("new").str)); "ok" }
  case "glob" => s"find . -name '${input("pat").str}'".!!.linesIterator.take(50).mkString("\n")
  case "grep" => s"grep -rn '${input("pat").str}' .".!!.linesIterator.take(50).mkString("\n")
  case "bash" => input("cmd").str.!!
).getOrElse("error")

val schema = ujson.read("""[{"name":"read","description":"Read","input_schema":{"type":"object","properties":{"path":{"type":"string"}},"required":["path"]}},
{"name":"write","description":"Write","input_schema":{"type":"object","properties":{"path":{"type":"string"},"content":{"type":"string"}},"required":["path","content"]}},
{"name":"edit","description":"Edit","input_schema":{"type":"object","properties":{"path":{"type":"string"},"old":{"type":"string"},"new":{"type":"string"}},"required":["path","old","new"]}},
{"name":"glob","description":"Find","input_schema":{"type":"object","properties":{"pat":{"type":"string"}},"required":["pat"]}},
{"name":"grep","description":"Search","input_schema":{"type":"object","properties":{"pat":{"type":"string"}},"required":["pat"]}},
{"name":"bash","description":"Run","input_schema":{"type":"object","properties":{"cmd":{"type":"string"}},"required":["cmd"]}}]""")

def ask(messages: ujson.Arr) = ujson.read(requests.post("https://api.anthropic.com/v1/messages",
  headers = Map("Content-Type" -> "application/json", "anthropic-version" -> "2023-06-01", "x-api-key" -> KEY),
  data = ujson.Obj("model" -> MODEL, "max_tokens" -> 4096, "system" -> "Concise assistant", "messages" -> messages, "tools" -> schema)).text())

@main def main() =
  println(s"${B}nanocode$R | $D$MODEL$R\n")
  var messages = ujson.Arr()
  
  while true do
    print(s"$B${BL}❯$R ")
    Option(StdIn.readLine()).map(_.trim) match
      case None | Some("/q") => return
      case Some("") => ()
      case Some("/c") => messages = ujson.Arr(); println(s"${G}⏺ Cleared$R")
      case Some(input) =>
        messages.value += ujson.Obj("role" -> "user", "content" -> input)
        var continue = true
        while continue do
          val resp = ask(messages)
          val content = resp("content").arr
          val results = ujson.Arr()
          content.foreach { block =>
            if block("type").str == "text" then println(s"\n$C⏺$R ${block("text").str}")
            if block("type").str == "tool_use" then
              val name = block("name").str
              println(s"\n$G⏺ $name$R")
              val result = runTool(name, block("input"))
              println(s"  $D⎿ ${result.linesIterator.nextOption.getOrElse("")}$R")
              results.value += ujson.Obj("type" -> "tool_result", "tool_use_id" -> block("id"), "content" -> result)
          }
          messages.value += ujson.Obj("role" -> "assistant", "content" -> content)
          if results.value.isEmpty then continue = false else messages.value += ujson.Obj("role" -> "user", "content" -> results)
        println()

File: functional/ocaml/nanocode.ml
(* nanocode - minimal claude code alternative (OCaml) *)
(* ocamlfind ocamlopt -package yojson,cohttp-lwt-unix,str -linkpkg nanocode.ml -o nanocode *)
open Lwt.Infix
open Str

let key =
  try Sys.getenv "OPENROUTER_API_KEY" with
  | Not_found -> Sys.getenv "ANTHROPIC_API_KEY"

let api =
  try Sys.getenv "API_URL" with
  | Not_found ->
      if Sys.getenv_opt "OPENROUTER_API_KEY" <> None then "https://openrouter.ai/api/v1/messages"
      else "https://api.anthropic.com/v1/messages"

let model =
  match Sys.getenv_opt "MODEL" with
  | Some m -> m
  | None ->
      if Sys.getenv_opt "OPENROUTER_API_KEY" <> None
      then "anthropic/claude-opus-4-5"
      else "claude-opus-4-5"
let r, b, d, c, g, bl = "\027[0m", "\027[1m", "\027[2m", "\027[36m", "\027[32m", "\027[34m"

let tool name input =
  let open Yojson.Basic.Util in
  match name with
  | "read" -> (try
      let lines = In_channel.with_open_text (input |> member "path" |> to_string) In_channel.input_all |> String.split_on_char '\n' in
      let off = input |> member "offset" |> to_int_option |> Option.value ~default:0 in
      let lim = input |> member "limit" |> to_int_option |> Option.value ~default:(List.length lines) in
      lines |> List.drop off |> fun l -> l |> List.take lim |> List.mapi (fun i l -> Printf.sprintf "%d| %s" (off+i+1) l) |> String.concat "\n"
    with _ -> "error")
  | "write" -> Out_channel.with_open_text (input |> member "path" |> to_string)
    (fun oc -> Out_channel.output_string oc (input |> member "content" |> to_string)); "ok"
  | "edit" -> (try
      let path = input |> member "path" |> to_string in
      let oldv = input |> member "old" |> to_string in
      let newv = input |> member "new" |> to_string in
      let all = input |> member "all" |> to_bool_option |> Option.value ~default:false in
      let txt = In_channel.with_open_text path In_channel.input_all in
      let occurrences = (split (regexp_string oldv) txt |> List.length) - 1 in
      if occurrences = 0 then "error: old_string not found"
      else if occurrences > 1 && not all then Printf.sprintf "error: old_string appears %d times, use all=true" occurrences
      else
        let updated =
          if all then Str.global_replace (Str.regexp_string oldv) newv txt
          else Str.replace_first (Str.regexp_string oldv) newv txt
        in
        Out_channel.with_open_text path (fun oc -> Out_channel.output_string oc updated); "ok"
    with _ -> "error")
  | "bash" -> let ic = Unix.open_process_in (input |> member "cmd" |> to_string) in
    let r = In_channel.input_all ic in ignore (Unix.close_process_in ic); r
  | "glob" -> let ic = Unix.open_process_in (Printf.sprintf "find . -name '%s' | head -50" (input |> member "pat" |> to_string)) in
    let r = In_channel.input_all ic in ignore (Unix.close_process_in ic); r
  | "grep" -> let ic = Unix.open_process_in (Printf.sprintf "grep -rn '%s' . | head -50" (input |> member "pat" |> to_string)) in
    let r = In_channel.input_all ic in ignore (Unix.close_process_in ic); r
  | _ -> "unknown"

let schema = {|[
{"name":"read","description":"Read","input_schema":{"type":"object","properties":{"path":{"type":"string"},"offset":{"type":"integer"},"limit":{"type":"integer"}},"required":["path"]}},
{"name":"write","description":"Write","input_schema":{"type":"object","properties":{"path":{"type":"string"},"content":{"type":"string"}},"required":["path","content"]}},
{"name":"edit","description":"Replace","input_schema":{"type":"object","properties":{"path":{"type":"string"},"old":{"type":"string"},"new":{"type":"string"},"all":{"type":"boolean"}},"required":["path","old","new"]}},
{"name":"bash","description":"Run","input_schema":{"type":"object","properties":{"cmd":{"type":"string"}},"required":["cmd"]}},
{"name":"glob","description":"Find","input_schema":{"type":"object","properties":{"pat":{"type":"string"},"path":{"type":"string"}},"required":["pat"]}},
{"name":"grep","description":"Search","input_schema":{"type":"object","properties":{"pat":{"type":"string"},"path":{"type":"string"}},"required":["pat"]}}
]|}

let ask messages =
  let body = `Assoc [
    "model", `String model; "max_tokens", `Int 4096; "system", `String "Concise assistant";
    "messages", messages; "tools", Yojson.Basic.from_string schema
  ] |> Yojson.Basic.to_string in
  let headers =
    let common = [ "Content-Type", "application/json"; "anthropic-version", "2023-06-01" ] in
    let auth =
      if Sys.getenv_opt "OPENROUTER_API_KEY" <> None
      then [ "Authorization", "Bearer " ^ key ]
      else [ "x-api-key", key ]
    in
    Cohttp.Header.of_list (common @ auth)
  in
  Cohttp_lwt_unix.Client.post ~headers ~body:(Cohttp_lwt.Body.of_string body)
    (Uri.of_string api)
  >>= fun (_, body) -> Cohttp_lwt.Body.to_string body >|= Yojson.Basic.from_string

let () =
  Printf.printf "%snanocode%s | %s%s%s\n\n" b r d model r;
  let messages = ref (`List []) in
  try while true do
    Printf.printf "%s%s❯%s " b bl r; flush stdout;
    let input = String.trim (read_line ()) in
    if input = "" then () else if input = "/q" then exit 0
    else if input = "/c" then (messages := `List []; Printf.printf "%s⏺ Cleared%s\n" g r)
    else begin
      let open Yojson.Basic.Util in
      messages := `List (to_list !messages @ [`Assoc ["role", `String "user"; "content", `String input]]);
      let rec loop () =
        let resp = Lwt_main.run (ask !messages) in
        let content = resp |> member "content" |> to_list in
        let results = List.filter_map (fun block ->
          match block |> member "type" |> to_string with
          | "text" -> Printf.printf "\n%s⏺%s %s" c r (block |> member "text" |> to_string); None
          | "tool_use" ->
            let name = block |> member "name" |> to_string in
            Printf.printf "\n%s⏺ %s%s\n" g name r;
            let result = tool name (block |> member "input") in
            Printf.printf "  %s⎿ %s%s\n" d (String.split_on_char '\n' result |> List.hd) r;
            Some (`Assoc ["type", `String "tool_result"; "tool_use_id", block |> member "id"; "content", `String result])
          | _ -> None
        ) content in
        messages := `List (to_list !messages @ [`Assoc ["role", `String "assistant"; "content", `List content]]);
        if results = [] then () else begin
          messages := `List (to_list !messages @ [`Assoc ["role", `String "user"; "content", `List results]]);
          loop ()
        end
      in loop (); print_newline ()
    end
  done with End_of_file -> ()
